oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/introspection/JwtPrincipalConverter.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright 2002-2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.server.resource.introspection;
+
+import org.springframework.security.oauth2.jwt.Jwt;
+
+/**
+ * @author Alex Vlasov
+ */
+public interface JwtPrincipalConverter {
+
+	Object convert(Jwt jwt, String principalName);
```

Comment:
Just connecting the two conversations here. This should return a Spring Security API. That allows Spring to evolve around the API over time and reason about its contents. If it returns `Object`, we lose that ability.

Reply:
For additional context, please see the code in [this sample application](https://github.com/spring-projects/spring-security-samples/blob/main/servlet/spring-boot/java/authentication/username-password/user-details-service/custom-user/src/main/java/example/CustomUserRepositoryUserDetailsService.java).

Reply:
Authentication.principal is an Object. Any other type would restrict principal type. If Object is not allowed by SS code convention, then could it be some marker interface? That marker interface could be a superinterface  for other principal interfaces like AuthenticatedPrincipal. Or maybe it could be AuthenticatedPrincipal itself. Sorry, I am not very familiar with SS architecture and not figure out what to do in this situation.

Reply:
No problem. What I'd recommend is that it return `OAuth2AuthenticatedPrincipal`. This allows a converter to transmit and user-level authorities (which can be different from the authorities derived from the JWT).

If implementing extra methods is a concern, I think we can look at adding `default` implementations to some of those methods.

Reply:
Current principal is `Jwt` object. It does not implement `OAuth2AuthenticatedPrincipal`. `JwtPrincipalConverter` should convert to some `Jwt` superclass to be compatible with current principal. What to do with this?